set permissions for the refactored controller

chr15p opened the pull request:
This will set up the various service accounts, roles and rolebindings needed by the new (refactored) controller.  These will all be applied at install (`make deploy`) time.
For clarity the current  manager setup (`config/rbac/role.yaml` and `role_binding.yaml`) are renamed, and each account is broken into its own set of files.

If applied before the new controller is made live the old controller will overwrite any changes here at reconcile time and continue to work (or at least will continue to be as broken as it is now...) so this PR can be applied safely without the refactor being completed.

marquiz commented on config/rbac/gc_clusterrole.yaml:
```diff
@@ -0,0 +1,32 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: gc
```
Should we use `nfd-` prefix for all the names. I think this is relevant especially for the cluster-scoped resources like this.

```suggestion
  name: nfd-gc
```

chr15p replied:
when `make deploy` runs kustomize it appends `nfd-` to all the names in the config it generates  (set by the `namePrefix: nfd-` setting in config/default/kustomization.yaml), so all the roles, service-accoutns etc should be named nfd-*

marquiz replied:
Hmm, ok. I still find it confusing that the clusterrole references a ServiceAccount with a different name than what's here 🤨 If the names are all replaced could we just use a placeholder (like `NAME`) for all of them to make it clear that it will be replaced. Or is it too obvious (to everyone else than me)?

yevgeny-shnaidman replied:
maybe we might remove the namePrefix? IMHO the definitions of the roles/accounts/etc' become more readable, we don't need to take all the substitutions into account when reading the yamls..

marquiz replied:
> IMHO the definitions of the roles/accounts/etc' become more readable, we don't need to take all the substitutions into account when reading the yamls..

I agree. Would make it more readable and easier to review and spot mistakes

chr15p replied:
I agree, it is a bit confusing as is, while testing it took me a bit to figure out why I was getting names like `nfd-nfd-gc`   I left it in to try and keep the focus of the PR tight, however it sounds like we all agree that its worth doing, so I will make the change and update the PR.

marquiz replied:
> I left it in to try and keep the focus of the PR tight,

Mm, that could also be done in a separate (prior) PR. If you like.

marquiz reviewed (commented):
/ok-to-test

yevgeny-shnaidman commented:
@chr15p WDYT about creating sub-directories (master, worker, prune etc') under config/rbac ? will make it a little bit clearer to understand which roles/sa relevant for which component

chr15p commented:
I've moved the components into their own subdirectories of rbac, and the parts needed for the operator itself (manager and leader_election role/sa etc)  into a `core` directory so hopefully its a little more obvious what files are used be what.

I have also removed the `namePrefix:` and hardcoded all the names to start with nfd-  to be less confusing. Everything in my test setup is working and labels are being applied by the new controller and all the object names look good so I think I've caught all the name changes.

Lastly I've synced the clusterroles/roles with the helm charts.

marquiz commented on config/manager/kustomization.yaml:
```diff
@@ -12,5 +12,5 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 - name: controller
-  newName: gcr.io/k8s-staging-nfd/node-feature-discovery-operator
-  newTag: master
+  newName: registry.k8s.io/nfd/node-feature-discovery-operator
+  newTag: 0.4.2
```
I'd suggest to leave this as it was

chr15p replied:
good catch. This gets set by the `make deploy` process so they keep creeping in while I'm testing, I obviously missed this one and committed it by mistake. I have reverted it.

marquiz reviewed (commented):
Wow @chr15p now this is so much easier for me to read 🚀 One small nit below but otherwise looks good

ArangoGutierrez reviewed (approved):
Approved, just a nit, could you quash the commits? after that I'll lgtm

chr15p commented:
Squished the commits and reworded the commit message to cover all the changes.

ArangoGutierrez reviewed (approved):
I'll leave the lgtm to @marquiz

marquiz reviewed (approved):
Good job on this @chr15p. A bit of extra work but I think it was worth it 😇
/lgtm
